docs/_build/man/mikado.1
.\" Man page generated from reStructuredText.
.
.TH "MIKADO" "1" "Aug 04, 2016" "1.0" "Mikado"
.SH NAME
mikado \- Mikado Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.sp
\fI\%python_badge\fP \fI\%snake_badge\fP
.INDENT 0.0
.TP
.B Authors
Venturini Luca,
Caim Shabhonam,
Mapleson Daniel,
Kaithakottil Gemy George,
Swarbreck David
.TP
.B Version
1.0 (July 2016)
.UNINDENT
.sp
Mikado is a lightweight Python3 pipeline to extract the best transcript models from multiple transcript assemblies.
.sp
Contents:
.SH INTRODUCTION
.sp
\fI\%python_badge\fP \fI\%snake_badge\fP
.sp
Introduction test
.SS Metrics
.sp
These are all the metrics available to quantify transcripts. The documentation for this section has been generated using the metrics utility\&.
.sp
Metrics belong to one of the following categories:
.INDENT 0.0
.IP \(bu 2
\fBDescriptive\fP: these metrics merely provide a description of the transcript (eg its ID) and are not used for scoring.
.IP \(bu 2
\fBcDNA\fP: these metrics refer to basic features of any transcript such as its number of exons, its cDNA length, etc.
.IP \(bu 2
\fBIntron\fP: these metrics refer to features related to the number of introns and their lengths.
.IP \(bu 2
\fBCDS\fP: these metrics refer to features related to the CDS assigned to the transcript.
.IP \(bu 2
\fBUTR\fP: these metrics refer to features related to the UTR of the transcript. In the case in which a transcript has been assigned multiple ORFs, unless otherwise stated the UTR metrics will be derived only considering the \fIselected\fP ORF, not the combination of all of them.
.IP \(bu 2
\fBLocus\fP: these metrics refer to features of the transcript in relationship to all other transcripts in its locus, eg how many of the introns present in the locus are present in the transcript. These metrics are calculated by Mikado during the picking phase, and as such their value can vary during the different stages as the transcripts are shifted to different groups.
.IP \(bu 2
\fBExternal\fP: these metrics are derived from accessory data that is recovered for the transcript during the run time. Examples include data regarding the number of introns confirmed by external programs such as PortCullis, or the BLAST score of the best hits.
.UNINDENT
.TS
center;
|l|l|l|l|.
_
T{
Metric name
T}	T{
Description
T}	T{
Data type
T}	T{
Category
T}
_
T{
\fItid\fP
T}	T{
Name of the transcript. Not used for scoring.
T}	T{
String
T}	T{
\fBDescriptive\fP
T}
_
T{
\fIparent\fP
T}	T{
Name of the transcript parent. Not used for scoring.
T}	T{
String
T}	T{
\fBDescriptive\fP
T}
_
T{
\fIscore\fP
T}	T{
Final score of the transcript.
T}	T{
Float
T}	T{
\fBDescriptive\fP
T}
_
T{
\fIbest_bits\fP
T}	T{
Best Bit Score associated with the transcript.
T}	T{
Float
T}	T{
\fBExternal\fP
T}
_
T{
\fIblast_score\fP
T}	T{
Alias for either \fIbest_bits\fP or \fIsnowy_blast_score\fP\&. Set
currently to \fIbest_bits\fP
T}	T{
Float
T}	T{
\fBExternal\fP
T}
_
T{
\fIcanonical_intron_proportion\fP
T}	T{
This metric returns the proportion of canonical introns of
the transcript on its total number of introns.
T}	T{
Float
Float
T}	T{
\fBIntron\fP
T}
_
T{
\fIcdna_length\fP
T}	T{
This property returns the length of the transcript.
T}	T{
Int
T}	T{
\fBcDNA\fP
T}
_
T{
\fIcds_not_maximal\fP
T}	T{
This property returns the length of the CDS excluding
that contained in the selected ORF. If the transcript only
has one ORF, this metric returns a value of 0.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIcds_not_maximal_fraction\fP
T}	T{
This property returns the fraction of bases not in the
selected ORF compared to the total number of CDS bases
in the cDNA.
T}	T{
Float
T}	T{
\fBCDS\fP
T}
_
T{
\fIcombined_cds_fraction\fP
T}	T{
This property return the percentage of the CDS part of the
transcript vs. the cDNA length.
T}	T{
Float
T}	T{
\fBCDS\fP
T}
_
T{
\fIcombined_cds_intron_fraction\fP
T}	T{
This property returns the fraction of CDS introns of the
transcript vs. the total number of CDS introns in the
Locus. If the transcript is by itself, it returns 1.
T}	T{
Float
T}	T{
\fBLocus\fP
T}
_
T{
\fIcombined_cds_length\fP
T}	T{
This property returns the fraction of CDS introns of the
transcript, across all its ORFs.
T}	T{
Float
T}	T{
\fBCDS\fP
T}
_
T{
\fIcombined_cds_num\fP
T}	T{
This property returns the number of non\-overlapping CDS
segments in the transcript.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIcombined_cds_num_fraction\fP
T}	T{
This property returns the fraction of non\-overlapping CDS
segments in the transcript vs. the total number of exons.
T}	T{
Float
T}	T{
\fBCDS\fP
T}
_
T{
\fIcombined_utr_fraction\fP
T}	T{
This property returns the fraction of the cDNA which is
not coding according to any ORF. Complement of
\fIcombined_cds_fraction\fP
T}	T{
Float
T}	T{
\fBUTR\fP
T}
_
T{
\fIcombined_utr_length\fP
T}	T{
This property return the length of the UTR part of the
transcript.
T}	T{
Int
T}	T{
\fBUTR\fP
T}
_
T{
\fIend_distance_from_junction\fP
T}	T{
This metric returns the cDNA distance between the stop
and the last junction of the transcript. In many
eukaryotes, this distance cannot exceed 50\-55 bps,
otherwise the transcript becomes a target for NMD. If the
transcript is not coding or there is no junction
downstream of the stop codon, the metric returns 0.
This metric considers the combined CDS end.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIend_distance_from_tes\fP
T}	T{
This property returns the distance of the end of the
combined CDS from the transcript end site. If no CDS is
defined, it defaults to 0.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIexon_fraction\fP
T}	T{
This property returns the fraction of exons of the
transcript which are contained in the sublocus. If the
transcript is by itself, it returns 1.
T}	T{
Float
T}	T{
\fBLocus\fP
T}
_
T{
\fIexon_num\fP
T}	T{
This property returns the number of exons of the
transcript.
T}	T{
Int
T}	T{
\fBcDNA\fP
T}
_
T{
\fIfive_utr_length\fP
T}	T{
Returns the length of the 5\(aq UTR of the \fIselected\fP ORF.
T}	T{
Int
T}	T{
\fBUTR\fP
T}
_
T{
\fIfive_utr_num\fP
T}	T{
This property returns the number of 5\(aq UTR segments for
the selected ORF.
T}	T{
Int
T}	T{
\fBUTR\fP
T}
_
T{
\fIfive_utr_num_complete\fP
T}	T{
This property returns the number of 5\(aq UTR segments for
the selected ORF, considering only those which are
complete exons.
T}	T{
Int
T}	T{
\fBUTR\fP
T}
_
T{
\fIhas_start_codon\fP
T}	T{
True if the selected ORF has a start codon, False
otherwise
T}	T{
Bool
T}	T{
\fBCDS\fP
T}
_
T{
\fIhas_stop_codon\fP
T}	T{
True if the selected ORF has a stop codon, False otherwise
T}	T{
Bool
T}	T{
\fBCDS\fP
T}
_
T{
\fIhighest_cds_exon_number\fP
T}	T{
This property returns the maximum number of CDS segments
among the ORFs; this number can refer to an ORF
\fIDIFFERENT\fP from the maximal ORF.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIhighest_cds_exons_num\fP
T}	T{
Returns the number of CDS segments in the selected ORF
(irrespective of the number of exons involved)
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIintron_fraction\fP
T}	T{
This property returns the fraction of introns of the
transcript vs. the total number of introns in the Locus.
If the transcript is by itself, it returns 1.
T}	T{
Float
T}	T{
\fBLocus\fP
T}
_
T{
\fIis_complete\fP
T}	T{
Boolean. True if the selected ORF has both start and end.
T}	T{
Bool
T}	T{
\fBCDS\fP
T}
_
T{
\fImax_intron_length\fP
T}	T{
This property returns the greatest intron length for the
transcript.
T}	T{
Int
T}	T{
\fBIntron\fP
T}
_
T{
\fImin_intron_length\fP
T}	T{
This property returns the smallest intron length for the
transcript.
T}	T{
Int
T}	T{
\fBIntron\fP
T}
_
T{
\fInon_verified_introns_num\fP
T}	T{
This metric returns the number of introns of the
transcript which are not validated by external data.
T}	T{
Int
T}	T{
\fBExternal\fP
T}
_
T{
\fInum_introns_greater_than_max\fP
T}	T{
This metric returns the number of introns greater than the
maximum acceptable intron size indicated in the
constructor.
T}	T{
Int
T}	T{
\fBIntron\fP
T}
_
T{
\fInum_introns_smaller_than_min\fP
T}	T{
This metric returns the number of introns smaller than the
mininum acceptable intron size indicated in the
constructor.
T}	T{
Int
T}	T{
\fBIntron\fP
T}
_
T{
\fInumber_internal_orfs\fP
T}	T{
This property returns the number of ORFs inside a
transcript.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIproportion_verified_introns\fP
T}	T{
This metric returns, as a fraction, how many of the
transcript introns are validated by external data.
T}	T{
Float
T}	T{
\fBExternal\fP
T}
_
T{
\fIproportion_verified_introns_inlocus\fP
T}	T{
This metric returns, as a fraction, how many of the
verified introns inside the Locus are contained inside the
transcript.
T}	T{
Float
T}	T{
\fBLocus\fP
T}
_
T{
\fIretained_fraction\fP
T}	T{
This property returns the fraction of the cDNA which is
contained in retained introns.
T}	T{
Float
T}	T{
\fBLocus\fP
T}
_
T{
\fIretained_intron_num\fP
T}	T{
This property records the number of introns in the
transcripts which are marked as being retained.
T}	T{
Int
T}	T{
\fBLocus\fP
T}
_
T{
\fIselected_cds_exons_fraction\fP
T}	T{
Returns the fraction of CDS segments in the selected ORF
(irrespective of the number of exons involved)
T}	T{
Float
T}	T{
\fBCDS\fP
T}
_
T{
\fIselected_cds_fraction\fP
T}	T{
This property calculates the fraction of the selected CDS
vs. the cDNA length.
T}	T{
Float
T}	T{
\fBCDS\fP
T}
_
T{
\fIselected_cds_intron_fraction\fP
T}	T{
This property returns the fraction of CDS introns of the
selected ORF of the transcript vs. the total number of
CDS introns in the Locus (considering only the selected
ORF). If the transcript is by itself, it should return 1.
T}	T{
Float
T}	T{
\fBCDS\fP
T}
_
T{
\fIselected_cds_length\fP
T}	T{
This property calculates the length of the CDS selected
as best inside the cDNA.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIselected_cds_num\fP
T}	T{
This property calculates the number of CDS exons for the
selected ORF.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIselected_cds_number_fraction\fP
T}	T{
This property returns the proportion of best possible CDS
segments vs. the number of exons. See selected_cds_number.
T}	T{
Float
T}	T{
\fBCDS\fP
T}
_
T{
\fIselected_end_distance_from_junction\fP
T}	T{
This metric returns the distance between the stop codon
and the last junction of the transcript. In many
eukaryotes, this distance cannot exceed 50\-55 bps,
otherwise the transcript becomes a target for NMD. If the
transcript is not coding or there is no junction
downstream of the stop codon, the metric returns 0.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIselected_end_distance_from_tes\fP
T}	T{
This property returns the distance of the end of the best
CDS from the transcript end site. If no CDS is defined,
it defaults to 0.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIselected_start_distance_from_tss\fP
T}	T{
This property returns the distance of the start of the
best CDS from the transcript start site. If no CDS is
defined, it defaults to 0.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIsnowy_blast_score\fP
T}	T{
Metric that indicates how good a hit is compared to the
competition, in terms of BLAST similarities. As in
SnowyOwl [SnowyOwl], the score for each hit is calculated
by taking the percentage of positive matches and dividing
it by (2 * len(self.blast_hits)). IMPORTANT: when
splitting transcripts by ORF, a blast hit is added to the
new transcript only if it is contained within it. This
will influnce directly this metric.
T}	T{
Float
T}	T{
\fBExternal\fP
T}
_
T{
\fIsource_score\fP
T}	T{
This metric returns a score that is assigned to the
transcript solely in virtue of its origin.
T}	T{
Float
T}	T{
\fBExternal\fP
T}
_
T{
\fIstart_distance_from_tss\fP
T}	T{
This property returns the distance of the start of the
combined CDS from the transcript start site.
If no CDS is defined, it defaults to 0.
T}	T{
Int
T}	T{
\fBCDS\fP
T}
_
T{
\fIthree_utr_length\fP
T}	T{
Returns the length of the 5\(aq UTR of the selected ORF.
T}	T{
Int
T}	T{
\fBUTR\fP
T}
_
T{
\fIthree_utr_num\fP
T}	T{
This property returns the number of 3\(aq UTR segments
(referred to the selected ORF).
T}	T{
Int
T}	T{
\fBUTR\fP
T}
_
T{
\fIthree_utr_num_complete\fP
T}	T{
This property returns the number of 3\(aq UTR segments for
the selected ORF, considering only those which are
complete exons.
T}	T{
Int
T}	T{
\fBUTR\fP
T}
_
T{
\fIutr_fraction\fP
T}	T{
This property calculates the length of the UTR of the
selected ORF vs. the cDNA length.
T}	T{
Float
T}	T{
\fBUTR\fP
T}
_
T{
\fIutr_length\fP
T}	T{
Returns the sum of the 5\(aq+3\(aq UTR lengths.
T}	T{
Int
T}	T{
\fBUTR\fP
T}
_
T{
\fIutr_num\fP
T}	T{
Returns the number of UTR segments.
T}	T{
Int
T}	T{
\fBUTR\fP
T}
_
T{
\fIutr_num_complete\fP
T}	T{
Returns the number of UTR segments which are complete
exons.
T}	T{
Int
T}	T{
\fBUTR\fP
T}
_
T{
\fIverified_introns_num\fP
T}	T{
This metric returns the number of introns of the
transcript which are validated by external data.
T}	T{
Int
T}	T{
\fBExternal\fP
T}
_
.TE
:: \fI\%Installation\fP
.SH INSTALLATION
.SS Python version
.sp
Mikado
.SS Download
.sp
Mikado is available on PyPI, so it is possible to install it with
.sp
\fBpip3 install mikado\fP
.sp
The source for the latest release can be obtained with
.sp
\fBpip3 download mikado\fP
.sp
As the package contains some core Cython components, it might be necessary to download and compile the source code instead of relying on the wheel.
.sp
Alternatively, Mikado can be installed from source by obtaining it from our \fI\%GitHub\fP repository. Either download the \fI\%latest release\fP  or download the latest stable development snapshot with
.sp
\fBgit clone https://github.com/lucventurini/mikado.git; cd mikado\fP
.sp
If you desire, the unstable development version can be obtained with the command
.sp
\fBgit checkout development\fP
.sp
in the Git directory. Please note that development can proceed quite rapidly.
.SS Building and installing from source
.sp
If you desire to install Mikado from source, this can be achieved with
.sp
\fBpython3 setup.py bdist_wheel\fP
.sp
Followed by
.sp
\fBpip3 install dist/*whl\fP
.sp
We advise to test whether the distribution has been built successfully by executing the unit test suite with
.sp
\fBpython3 setup.py test\fP
.sp
Although code coverage is not perfect yet, it is over 50% for the whole package and considerably higher for the core components.
.SS Python Dependencies
.sp
Mikado has been written for Python 3.4 and 3.5. It is dependent on the following Python3 modules:
.INDENT 0.0
.IP \(bu 2
wheel>=0.28.0
.IP \(bu 2
pyyaml [PyYaml]
.IP \(bu 2
jsonschema
.IP \(bu 2
Cython [Cython]
.IP \(bu 2
cython [Cython]
.IP \(bu 2
numpy [Numpy]
.IP \(bu 2
networkx>=1.10 [NetworkX]
.IP \(bu 2
sqlalchemy>=1
.IP \(bu 2
sqlalchemy_utils
.IP \(bu 2
biopython>=1.66 [BioPython]
.IP \(bu 2
intervaltree
.IP \(bu 2
nose
.IP \(bu 2
pyfaidx
.IP \(bu 2
scikit\-learn>=0.17.0 [SciKit]
.IP \(bu 2
scipy>=0.15.0 [Scipy]
.IP \(bu 2
frozendict
.IP \(bu 2
python\-magic
.IP \(bu 2
drmaa [DRMAA]
.IP \(bu 2
snakemake [Snake]
.IP \(bu 2
docutils
.UNINDENT
.sp
These dependencies will be installed automatically by PIP.
.SS Additional dependencies
.sp
Mikado relies on relational databases for its functioning, so one of SQLite, PosGRESql or MySQL has to present for if to function properly. Additionally, the 
.nf
Daijin_
.fi
 pipeline requires BLAST+ and TransDecoder for the Mikado stage, and at least one RNA\-Seq aligner and one assembler, to be installed. If you are planning to execute it on a cluster, we do support job submission on SLURM, LSF and PBS clusters, either in the presence or absence of DRMAA.
.SH TUTORIAL
.SH USAGE
.sp
Mikado is composed of four different programs (\fIconfigure, prepare, serialise, pick\fP) which have to be executed serially to go from an ensemble of different assemblies to the final dataset. In addition to these core programs, Mikado provides a utility to compare annotations, similarly to CuffCompare and ParsEval (\fIcompare\fP), and various other minor utilities to perform operations such as extracting regions from a GFF, convert between different gene annotation formats, etc.
.SH MIKADO PIPELINE STAGES
.SS Mikado configure
.sp
This utility prepares the configuration file that will be used throughout the pipeline stages. While the most important options can be set at runtime through the command line, many algorithmic details can be accessed and intervened upon only through the file produced through this command.
.SS Usage
.sp
This command will generate a configuration file (in either JSON or YAML format), with the correct configuration for the parameters set on the command line. See \fI\%the in\-depth section on the structure of the configuration file\fP for details.
.sp
Command line parameters:
.INDENT 0.0
.IP \(bu 2
\fIfull\fP: By default, Mikado configure will output a stripped\-down configuration file, with only some of the fields explicitly present. Use this flag to show all the available fields.
.UNINDENT
.sp
\fBHINT:\fP
.INDENT 0.0
.INDENT 3.5
If a parameter is not explicitly present, its value will be set to the available default.
.UNINDENT
.UNINDENT
.INDENT 0.0
.IP \(bu 2
\fIexternal\fP: an external JSON/YAML file from which Mikado should recover parameters. If something is specified both on the command line and in the external configuration file, \fBcommand line options take precedence\fP\&.
.IP \(bu 2
\fIreference\fP: Reference genome FASTA file. Required for the correct functioning of Mikado.
.IP \(bu 2
\fIgff\fP: list of GFFs/GTFs of the assemblies to use for Mikado, comma separated.
.IP \(bu 2
\fIlabels\fP: optional list of labels to be assigned to each assembly. The label will be also the "source" field in the output of Mikado prepare\&.
.IP \(bu 2
\fIstrand\-specific\-assemblies\fP: list of strand specific assemblies among those specified with the \fIgff\fP flag.
.IP \(bu 2
\fIstrand\-specific\fP: flag. If set, all assemblies will be treated as strand\-specific.
.IP \(bu 2
.INDENT 2.0
.TP
.B \fIlist\fP: in alternative to specifying all the information on the command line, it is possible to give to Mikado a \fItab\-separated\fP file with the following contents:
.INDENT 7.0
.IP 1. 3
Location of the file
.IP 2. 3
label for the file
.IP 3. 3
whether that assembly is strand\-specific or not (write True/False)
.IP 4. 3
Optionally, a bonus/malus to be associated to transcripts coming from that assembly.
.UNINDENT
.UNINDENT
.IP \(bu 2
\fI\-j\fP, \fIjson\fP: flag. If present, the output file will be in JSON rather that YAML format.
.IP \(bu 2
\fImode\fP: in which mode Mikado pick will run. See the \fI\%relative section\fP for further details.
.IP \(bu 2
.INDENT 2.0
.TP
.B \fIscoring\fP: scoring file to be used for selection. At the moment, four configuration files are present:
.INDENT 7.0
.IP \(bu 2
plants
.IP \(bu 2
human
.IP \(bu 2
worm
.IP \(bu 2
insects
.UNINDENT
.UNINDENT
.IP \(bu 2
\fIcopy\-scoring\fP: this flag specifies a file the scoring file will be copied to. Useful for customisation on new species. The configuration file will account for this and list the location of the \fIcopied\fP file as the scoring file to be used for the run.
.IP \(bu 2
\fIjunctions\fP: Reliable junctions derived from the RNA\-Seq alignments. Usually obtained through \fI\%Portcullis\fP, but we also support eg junctions from Tophat2. This file must be in \fI\%BED12\fP format.
.IP \(bu 2
\fIblast_targets\fP: FASTA database of blast targets, for the serialisation stage.
.UNINDENT
.sp
Usage:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ mikado configure \-\-help
usage: Mikado configure [\-h] [\-\-full]
                        [\-\-scoring {insects.yaml,human.yaml,plants.yaml,worm.yaml}]
                        [\-\-copy\-scoring COPY_SCORING] [\-\-strand\-specific]
                        [\-\-no\-files | \-\-gff GFF | \-\-list LIST]
                        [\-\-reference REFERENCE] [\-\-junctions JUNCTIONS]
                        [\-bt BLAST_TARGETS]
                        [\-\-strand\-specific\-assemblies STRAND_SPECIFIC_ASSEMBLIES]
                        [\-\-labels LABELS] [\-\-external EXTERNAL]
                        [\-\-mode {nosplit,stringent,lenient,permissive,split}]
                        [\-j]
                        [out]

This utility guides the user through the process of creating a configuration file for Mikado.

positional arguments:
  out

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-\-full
  \-\-strand\-specific     Boolean flag indicating whether all the assemblies are strand\-specific.
  \-\-no\-files            Remove all files\-specific options from the printed configuration file.
                                               Invoking the "\-\-gff" option will disable this flag.
  \-\-gff GFF             Input GFF/GTF file(s), separated by comma
  \-\-list LIST           List of the inputs, one by line, in the form:
                        <file1>  <label>  <strandedness (true/false)>
  \-\-reference REFERENCE
                        Fasta genomic reference.
  \-\-strand\-specific\-assemblies STRAND_SPECIFIC_ASSEMBLIES
                        List of strand\-specific assemblies among the inputs.
  \-\-labels LABELS       Labels to attach to the IDs of the transcripts of the input files,
                                separated by comma.
  \-\-external EXTERNAL   External configuration file to overwrite/add values from.
                            Parameters specified on the command line will take precedence over those present in the configuration file.
  \-\-mode {nosplit,stringent,lenient,permissive,split}
                        Mode in which Mikado will treat transcripts with multiple ORFs.
                        \- nosplit: keep the transcripts whole.
                        \- stringent: split multi\-orf transcripts if two consecutive ORFs have both BLAST hits
                                     and none of those hits is against the same target.
                        \- lenient: split multi\-orf transcripts as in stringent, and additionally, also when
                                   either of the ORFs lacks a BLAST hit (but not both).
                        \- permissive: like lenient, but also split when both ORFs lack BLAST hits
                        \- split: split multi\-orf transcripts regardless of what BLAST data is available.
  \-j, \-\-json            Output will be in JSON instead of YAML format.

Options related to the scoring system:
  \-\-scoring {insects.yaml,human.yaml,plants.yaml,worm.yaml}
                        Available scoring files.
  \-\-copy\-scoring COPY_SCORING
                        File into which to copy the selected scoring file, for modification.

Options related to the serialisation step:
  \-\-junctions JUNCTIONS
  \-bt BLAST_TARGETS, \-\-blast_targets BLAST_TARGETS
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Anatomy of the configuration file
.sp
The guide here describes all voices of the configuration file. However, the configuration created by default by \fBmikado configure\fP is much simpler
.SS Database settings
.sp
This section deals with the database settings that will be necessary for the serialisation and picking phases of the pipeline. By default, Mikado will use a \fI\%SQLite database\fP, but it currently also supports \fI\%MySQL\fP and \fI\%PostgreSQL\fP through \fI\%SQLAlchemy\fP\&. Fields:
.INDENT 0.0
.IP \(bu 2
db: name of the database to use. In case the database is SQLite, this will be the database file, otherwise it will be the database \fIname\fP\&.
.IP \(bu 2
dbtype: one of:
* sqlite
* mysql
* postgresql
.IP \(bu 2
dbhost: host where the database is located. \fBRequired with MySQL and PostgreSQL\fP\&.
.IP \(bu 2
dbuser: User of the database. \fBRequired with MySQL and PostgreSQL\fP\&.
.IP \(bu 2
dbpasswd: Database password. \fBRequired with MySQL and PostgreSQL\fP\&.
.IP \(bu 2
dbport: Port to access to the database. It defaults to the normal ports for the selected database.
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
db_settings:
  #  Settings related to DB connection. Parameters:
  #  db: the DB to connect to. Required. Default: mikado.db
  #  dbtype: Type of DB to use. Choices: sqlite, postgresql, mysql. Default: sqlite.
  #  dbhost: Host of the database. Unused if dbtype is sqlite. Default: localhost
  #  dbuser: DB user. Default:
  #  dbpasswd: DB password for the user. Default:
  #  dbport: Integer. It indicates the default port for the DB.
  db: /usr/users/ga002/venturil/workspace/mikado/docs/mikado.db
  dbhost: localhost
  dbpasswd: \(aq\(aq
  dbport: 0
  dbtype: sqlite
  dbuser: \(aq\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Reference settings
.sp
This section of the configuration file deals with the reference genome. It specifies two fields:
.INDENT 0.0
.IP \(bu 2
genome: the genome FASTA file. \fBRequired\fP\&.
.IP \(bu 2
genome_fai: FAI index of the genome. Used by Mikado serialise, it can be inferred if left null.
.IP \(bu 2
transcriptome: optional annotation file for the genome. Mikado currently ignores this field, but it is used by Daijin to guide some of the RNA\-Seq assemblies.
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
reference:
  #  Options related to the reference genome.
  genome: \(aq\(aq
  genome_fai: \(aq\(aq
  transcriptome: \(aq\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Settings for the prepare stage
.sp
This section of the configuration file deals with the prepare stage of Mikado\&. It specifies the input files, their labels, and which of them are strand specific. The available fields are the following:
.INDENT 0.0
.IP \(bu 2
canonical: this voice specifies the splice site donors and acceptors that are considered canonical for the species. By default, Mikado uses the canonical splice site (GT/AG) and the two semi\-canonical pairs (GC/AG and AT/AC). Type: Array of two\-element arrays, composed by two\-letter strings.
.IP \(bu 2
lenient: boolean value. If set to \fIfalse\fP, transcripts that only have non\-canonical splice sites will be \fBremoved\fP from the output.
.IP \(bu 2
minimum_length: minimum length of the transcripts to be kept.
.IP \(bu 2
procs: number of processors to be used.
.IP \(bu 2
strand_specific: boolean. If set to \fItrue\fP, \fBall\fP input assemblies will be treated as strand\-specific, therefore keeping the strand of monoexonic fragments as it was.
.IP \(bu 2
strip_cds: boolean. If set to \fItrue\fP, the CDS features will be stripped off the input transcripts. This might be necessary for eg transcripts obtained through alignment with \fI\%GMAP\fP [GMAP]\&.
.IP \(bu 2
.INDENT 2.0
.TP
.B files: this sub\-section is the most important, as it contains among other things the locations and labels for the input files. Voices:
.INDENT 7.0
.IP \(bu 2
gff: array of the input files, in GFF or GTF format. Please note that only CDS/exon/UTR features will be considered from these files.
.IP \(bu 2
labels: optional array of the labels to be assigned to the input files. If non\-empty, \fIit must be of the same order and length of the gff array\fP, and be composed of unique elements. The labels will be used in two ways:
* as a prefix of the transcripts coming from the corresponding GFF
* as the \fIsource field\fP assigned to the transcript. This might be of relevance \fI\%during the picking stage\fP\&.
.IP \(bu 2
log: name of the log file.
.IP \(bu 2
out: name of the GTF output file.
.IP \(bu 2
out_fasta: name of the corresponding output FASTA file.
.IP \(bu 2
output_dir: output directory. It will be created if it does not exist already.
.IP \(bu 2
strand_specific_assemblies: array of the names of the GFF/GTF files that are strand specific. \fBAll the file names in this array must also appear in the gff array as well.\fP\&.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
prepare:
  #  Options related to the input data preparation.
  #  \- files: options relative to the input/output files.
  #  \- procs: Number of processes to use.
  #  \- strip_cds: whether to remove the CDS from the predictions during preparation.
  #  \- lenient: if set to True, invalid transcripts will be only flagged and not removed.
  #  EXPERIMENTAL.
  #  \- strand_specific: if set to True, transcripts will be assumed to be in the correct
  #  orientation, no strand flipping or removal
  #  \- strand_specific_assemblies: array of input predictions which are to be considered
  #  as strand\-specific.
  #    Predictions not in this list will be considered as non\-strand\-specific.
  #  \- canonical: canonical splice sites, to infer the correct orientation.
  canonical:
  \- \- GT
    \- AG
  \- \- GC
    \- AG
  \- \- AT
    \- AC
  files:
    #  Options related to the input and output files.
    #  \- out: output GTF file
    #  \- out_fasta: output transcript FASTA file
    #  \- gff: array of input predictions for this step.
    #  \- log: output log. Default: prepare.log
    #  \- labels: labels to be associated with the input GFFs. Default: None.
    gff: []
    labels: []
    log: prepare.log
    out: mikado_prepared.gtf
    out_fasta: mikado_prepared.fasta
    output_dir: .
    strand_specific_assemblies: []
  lenient: false
  minimum_length: 200
  procs: 1
  single: false
  strand_specific: false
  strip_cds: false
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Settings for the serialisation stage
.sp
This section of the configuration file deals with the serialisation stage of Mikado\&. It specifies the location of the ORF BED12 files from TransDecoder, the location of the XML files from BLAST, the location of portcullis junctions, and other details important at run time. It has the following voices:
.INDENT 0.0
.IP \(bu 2

util/mikado.py
#!/usr/bin/env python3

import argparse
import mikado_lib.subprograms


def all():
    """
    Mock. It will launch the whole pipeline.
    """
    parser = argparse.ArgumentParser()
    return parser


def main():

    """
    Main launcher function for the pipeline.
    """

    parser = argparse.ArgumentParser(prog="mikado",
                                     description="""Mikado is a program to analyse RNA-Seq data and determine the best
                                     transcript for each locus in accordance to user-specified criteria.""")
    subparsers = parser.add_subparsers(
        title="Components",
        help="These are the various components of Mikado."
    )

    subparsers.add_parser("prepare", description="GTF preparation script",
                          help="""Mikado prepare analyses an input GTF file and prepares it for the picking analysis
                          by sorting its transcripts and performing some simple consistency checks.""")
    subparsers.choices["prepare"] = mikado_lib.subprograms.prepare.prepare_parser()
    subparsers.choices["prepare"].prog = "mikado prepare"

    subparsers.add_parser("serialise", description="Data serialisation script",
                          help="""Mikado serialise creates the database used by the pick program. It handles
                          Junction and ORF BED12 files and BLAST XML results.""")
    subparsers.choices["serialise"] = mikado_lib.subprograms.serialise.serialise_parser()
    subparsers.choices["serialise"].prog = "mikado serialise"

    subparsers.add_parser("pick", description="Comparison script",
                          help="""Mikado pick analyses a sorted GTF/GFF files in order to identify its loci and choose
                          the best transcripts according to user-specified criteria.
                          It is dependent on files produced by the "prepare" and "serialise" components.
                          """)
    subparsers.choices["pick"] = mikado_lib.subprograms.pick.pick_parser()
    subparsers.choices["pick"].prog = "mikado pick"

    subparsers.add_parser( "compare", description="Comparison between reference and prediction",
                           help="""Mikado compare produces a detailed comparison of reference and prediction files.
                           It has been directly inspired by Cufflinks' cuffcompare and ParsEval.""")
    subparsers.choices["compare"] = mikado_lib.subprograms.compare.compare_parser()
    subparsers.choices["compare"].prog = "mikado compare"

    subparsers.add_parser("util", description = "Miscellaneous utilities",
                          help="Subparser holding various utilities of the suite.")
    subparsers.choices["util"] = mikado_lib.subprograms.util.util_parser()
    subparsers.choices["util"].prog = "mikado util"

    args = parser.parse_args()
    if hasattr(args, "func"):
        args.func(args)
    else:
        parser.print_usage()

if __name__ == '__main__':
    main()
